src/test/java/test/dataprovider/DataProviderRerunListener.java

```diff
@@ -0,0 +1,22 @@
+package test.dataprovider;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+
+import org.testng.IAnnotationTransformer;
+import org.testng.IRetryAnalyzer;
+import org.testng.annotations.ITestAnnotation;
+
+public class DataProviderRerunListener implements IAnnotationTransformer {
```

Comment:
Is this listener required ? The test case for sure needs a retry analyzer. So why not directly set it via the `@Test` annotation itself, instead of depending on a listener for wiring it in?